blackjax/adaptation/mclmc_adaptation.py

```diff
@@ -0,0 +1,243 @@
+
+    def step(iteration_state, weight_and_key):
+        outer_weight, rng_key = weight_and_key
+        """does one step of the dynamics and updates the estimate of the posterior size and optimal stepsize"""
+        state, params, adaptive_state, kalman_state = iteration_state
+        state, params, params_final, adaptive_state, success = predictor(
+            state, params, adaptive_state, rng_key
+        )
+        position, _ = ravel_pytree(state.position)
+        kalman_state = update_kalman(
+            position, kalman_state, outer_weight, success, params.step_size
+        )
+
+        return (state, params_final, adaptive_state, kalman_state), None
+
+    def L_step_size_adaptation(state, params, num_steps, rng_key):
+        num_steps1, num_steps2 = int(num_steps * frac_tune1), int(
+            num_steps * frac_tune2
+        )
+        # L_step_size_adaptation_keys = jax.random.split(rng_key, num_steps1 + num_steps2)
+        L_step_size_adaptation_keys = jnp.array([rng_key] * (num_steps1 + num_steps2))
+
+        # we use the last num_steps2 to compute the diagonal preconditioner
+        outer_weights = jnp.concatenate((jnp.zeros(num_steps1), jnp.ones(num_steps2)))
+
+        # initial state of the kalman filter
+        kalman_state = (0.0, jnp.zeros(dim), jnp.zeros(dim))
+
+        # run the steps
+        kalman_state = jax.lax.scan(
+            step,
+            init=(state, params, adap0, kalman_state),
+            xs=(outer_weights, L_step_size_adaptation_keys),
+            length=num_steps1 + num_steps2,
+        )[0]
+        state, params, _, kalman_state_output = kalman_state
+
+        L = params.L
+        # determine L
+        if num_steps2 != 0.0:
+            _, F1, F2 = kalman_state_output
+            variances = F2 - jnp.square(F1)
+            L = jnp.sqrt(jnp.sum(variances))
+
+        return state, MCLMCAdaptationState(L, params.step_size)
+
+    return L_step_size_adaptation
+
+
+def make_adaptation_L(kernel, frac, Lfactor):
+    """determine L by the autocorrelations (around 10 effective samples are needed for this to be accurate)"""
+
+    def adaptation_L(state, params, num_steps, key):
+        num_steps = int(num_steps * frac)
+        # adaptation_L_keys = jax.random.split(key, num_steps)
+        adaptation_L_keys = jnp.array([key] * (num_steps))
+
+        # run kernel in the normal way
+        state, info = jax.lax.scan(
+            f=lambda s, k: (
+                kernel(rng_key=k, state=s, L=params.L, step_size=params.step_size)
+            ),
+            init=state,
+            xs=adaptation_L_keys,
+        )
+        samples = info.transformed_position  # tranform is the identity here
+        flat_samples, _ = ravel_pytree(samples)
+        dim = pytree_size(state.position)
+        flat_samples = flat_samples.reshape(-1, dim)
+        ESS = 0.5 * effective_sample_size(
+            jnp.array([flat_samples, flat_samples])
+        )  # TODO: should only use a single chain here
```

Comment:
```suggestion
        _, unravel_fn = ravel_pytree(state.position)
        flat_samples = jax.vmap(unravel_fn)(samples)
        flat_samples = flat_samples.reshape(2, num_steps // 2, -1)
        ESS = effective_sample_size(flat_samples)
```

Reply:
Ah right... try:
```python
                flat_samples = jax.vmap(lambda x: ravel_pytree(x)[0])(samples)
	        flat_samples = flat_samples.reshape(2, num_steps // 2, -1)
	        ESS = effective_sample_size(flat_samples)
```

Reply:
This passes the test, but gives a numerically different result (slightly) such that the original MCLMC and blackjax implementation no longer align. Is there a reason effective_sample_size forbids a single chain?

Reply:
Because `effective_sample_size` by design use the within chain and cross chain variance to compute ess, so most implementation would split the single chain into 2 (see https://mc-stan.org/docs/2_18/reference-manual/effective-sample-size-section.html). 
I think it is fine if we dont align here, because this is a more accurate implementation IMO.

Reply:
OK, updated.